### DISCUSSION TITLE
The RAG don't work !

### DISCUSSION THREAD

**Discussion by Domi31tls** [open-webui/open-webui#5460] [General]
# Bug Report
When I ask a question about a document or web,  the response is always negative; the system does not see the document.

## Installation Method

Ubuntu 24.04
Docker
Version open-webui v0.3.21
Ollama
Mistal-nemo

[Describe the method you used to install the project, e.g., git clone, Docker, pip, etc.]

Docker pull

## Environment

Version open-webui v0.3.21
Version 0.3.9 local
Ubuntu 24.04

**Confirmation:**

- [ ] I have read and followed all the instructions provided in the README.md.
- [ ] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ ] I have included the Docker container logs.
- [ ] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

[Describe what you expected to happen.]

## Actual Behavior:

[Describe what actually happened.]

No reponse when I ask about a document or web adress :

## Description

**Bug Summary:**
Regardless of the question about a document or a website, the system does not see the document. I have tried with other embedded systems and encountered the same issue. Of course, the problem is the same when I load a page with #. The tests I am conducting are on text documents that are only 2 pages long.

**Docker Container Logs:**
No errors or warning

**Screenshots/Screen Recordings (if applicable):**

![image](https://github.com/user-attachments/assets/418a88a2-b2d0-4230-aa92-5250241cfe9c)
![image](https://github.com/user-attachments/assets/0c9a7071-2aff-4f69-8c5e-1f6b8b1050e2)

**Comment by Domi31tls** [comment]
I usee this RAG model : Use the following context as your learned knowledge, inside <context></context> XML tags.
<context>
    [context]
</context>

When answer to user:
- If you don't know, just say that you don't know.
- If you don't know when you are not sure, ask for clarification.
Avoid mentioning that you obtained the information from the context.
And answer according to the language of the user's question.

Given the context information, answer the query.
Query: [query]

**Comment by thiswillbeyourgithub** [comment]
I fixed it already, it will be in the next release. #5378 
You can try the dev branch or wait for the next release. Maybe try disabling hybrid search too if you set it.


Edit: 
See #5323 for why I don't recommend the model you're using
It might also be related to the lack of extension of the file you gave. Try with it ending by .txt ?

**Reply by Domi31tls** [reply]
thank you

**Reply by qiulang** [reply]
@thiswillbeyourgithub I use the latest dev code, which contains your fix to see whether RAG works now. 

I don't know whether I use it correctly or not, I find there is no difference and RAG still fails to work in my case. I have to use # to specify then exact document the RAG can give me the answer, otherwise RAG's answer is just nonsense. 

I also try to add and remove documents to my model's **Knowledge**, but it makes no difference. 

What I said in https://github.com/open-webui/open-webui/discussions/3868 seems to the case still, [queryDoc](https://github.com/open-webui/open-webui/blob/main/src/lib/apis/rag/index.ts#L299C14-L299C22) calls the RAG api in backend. But I don't know Svelte so I can't find out how queryDoc is called. When I search queryDoc in src folder, I can't find any reference to it.

**Reply by silentoplayz** [reply]
@qiulang Have you reset your vector storage within Open WebUI after changing the embedding model it uses? This is necessary and is even mentioned as a tip within the UI itself. Your documents will also need to be removed and re-added.

**Reply by qiulang** [reply]
I didn't need to reset my vector as I use the default embedding model. I just want to check if RAG can work.

So what I did is upload my documents into docker (`./app/backend/data/docs`),  in **Admin Panel**  -> **Documents**, click scan button, confirm from the log, all documents were scanned.

Then asked some question I knew answer lied in one of documents, but RAG gave me nonsense.

Then editted my model, add the documents to the model's **Knowledge**, asked the question again, still the answer is nonsense,

Then using # to specify the document and asked the question again, this time RAG works.

So from my test, the patch didn't work.

**Reply by thiswillbeyourgithub** [reply]
I think in the log you can see the rag template being replaced. This way you could see if it's empty content or not. Have you tried fiddling with the embedder parameter like topk etc ?

Also ask simple stuff like "what did you receive in your content" and set the threshold to 0

**Reply by qiulang** [reply]
From the log I can see the scan result like "[open_webui.apps.rag.main] store_docs_in_vector_db [Document(metadata={'source': '/app/backend/data/docs/..."  , but how do I check if  the rag template is replaced ?

**Reply by thiswillbeyourgithub** [reply]
Look for section of text that are part of the rag template

**Reply by qiulang** [reply]
@thiswillbeyourgithub no matter how I tried it just won't work and I almost feel like to give up. Maybe you can  send me your documents, your question and your generated RAG answer ?

What I test is rather simple. In one of my tests, I first get some pdfs from these news:

1. https://apnews.com/article/telegram-pavel-durov-arrest-2c8015c102cce23c23d55c6ca82641c5
2. https://edition.cnn.com/2024/08/28/tech/pavel-durov-telegram-custody-released-intl/index.html
3. https://www.nytimes.com/2024/09/05/technology/telegram-pavel-durov-comments.html

Let them scanned and just asked what happen to Telegram’s Founder in paris

**Reply by thiswillbeyourgithub** [reply]
To pinpoint the issue you should do a simpler minimal test. Create a .txt file containing only "the answer is 'ratatouille'" and use rag with thr query "what's the answer".

Then you should try sending a pdf and ask " what content did I sent you" as docs are called content in the default rag template.

Finally of you could setup langfuse that would help you a lot because you'd be able to see the exact prompt asked to the LLM. Alternatively use a openwebui filter that prints the prompt in the inlet and read the logs

**Reply by qiulang** [reply]
What I said was that since you believed you had fixed the problem, can you send me your documents, your question and your generated RAG answer ? Then I can verify from my test.

**Reply by thiswillbeyourgithub** [reply]
Sorry I'm very short on time and can't do that. But there's really no need for testing RAG, you should instead folliw the steps in my previous message IMO

**Reply by qiulang** [reply]
Well, to be honest with you I almost decide to give it up. I have tested another open source RAG project https://github.com/1Panel-dev/MaxKB and it just works (at least in my case) while no matter how I try open-webui, it just does not work. Why bother then.

**Reply by silentoplayz** [reply]
> I didn't need to reset my vector as I use the default embedding model. I just want to check if RAG can work.
> 
> So what I did is upload my documents into docker (`./app/backend/data/docs`), in **Admin Panel** -> **Documents**, click scan button, confirm from the log, all documents were scanned.
> 
> Then asked some question I knew answer lied in one of documents, but RAG gave me nonsense.
> 
> Then editted my model, add the documents to the model's **Knowledge**, asked the question again, still the answer is nonsense,
> 
> Then using # to specify the document and asked the question again, this time RAG works.
> 
> So from my test, the patch didn't work.

@qiulang None of what you did will get to the root cause of your issue if your embeddings doesn’t match the model that generated said embeddings for your documents. I advise you to try what I’ve mentioned previously to see if this solves your issue. You may simply need to reset your vector database, remove your documents from the workspace, and re-add them with the scan button to generate new embeddings. What version of Open WebUI are you running? Can you share your `Documents` settings in Open WebUI via a screenshot? A picture is worth a thousand words and someone may be able to spot anything out of the ordinary pretty quickly if something were wrong.

**Reply by qiulang** [reply]
@silentoplayz Actually I have done all of that. I compare 6 embedding models and I find `BAAI/bge-base-zh-v1.5` performance best  for Chinese language. For English, I find "BAAI/bge-small-en-v1.5" is better than "all-MiniLM-L12-v2"

```
# model = SentenceTransformer("all-MiniLM-L12-v2")
# model = SentenceTransformer("BAAI/bge-small-en-v1.5")
# model = SentenceTransformer("distiluse-base-multilingual-cased-v1")
# model = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")
# model = SentenceTransformer("BAAI/bge-small-zh-v1.5")
model = SentenceTransformer("BAAI/bge-base-zh-v1.5")
```

I compare qwen2:7b against llama3.1:latest 

I have tested 0.3.21 ,0.3.16. 0.3.18 and the latest dev branch. And of course for every version I reset database.

I open an issue but the issue was turn into another discussion that never got response.

So basically I gave up.

**Reply by thiswillbeyourgithub** [reply]
Shameless plug: i personnaly made my own very advanced RAG project there at [wdoc](https://github.com/thiswillbeyourgithub/WDoc)

It has no gui and probably has bugs that I'll fix only after my exams but it might still be of interest.

My medium term plan is to integrate wdoc with open webui via pipelines or something like that.

**Reply by qiulang** [reply]
@thiswillbeyourgithub I will try yours later. BTW do you have plan to use llamaindex ?

**Reply by thiswillbeyourgithub** [reply]
Well I don't know where llamaindex would fit into it. It's my own lib based on langchain but it doesn't really matter as I was not satisfied with any of those native RAG setup. If you have anything i'm missing to suggest don't hesitate to open up an issue over there.

**Reply by silentoplayz** [reply]
Related - https://github.com/open-webui/open-webui/discussions/3868#discussioncomment-10782274

**Comment by dmvieira** [comment]
I did a new feature request related with this issue: https://github.com/open-webui/open-webui/issues/5495

**Comment by Domi31tls** [comment]
For me, RAG wasn't working at first. But after testing several settings, I now get good results with txt files. I'm using Docker Open-WebUI in its v0.3.21 version on Linux with Ollama, all locally. You can see my conf:
![image](https://github.com/user-attachments/assets/a90351a4-dc4e-400b-a16c-8e8de0fa7e1c)

**Comment by silentoplayz** [comment]
Sharing my RAG settings for those who may want to copy it.
![image](https://github.com/user-attachments/assets/b6423019-df62-4b2c-9f19-3789d64a8d58)

The RAG template I use:
```
You are to use the following context as your learned knowledge, inside the <context></context> XML tags.
<KnowledgeGraph>
<context>
<!-- Use this context as your learned knowledge -->
[context]
</context>
</KnowledgeGraph>

<dialog_management>
* UncertaintyPrinciple: 
    + If uncertain or lacking information, acknowledge the limitation and ask for clarification or context.
    + Refrain from providing speculative or potentially misleading answers.
* LanguageAdaptation: 
    + Respond using the user's language, maintaining cultural and regional nuances.
    + Be sensitive to tone, idioms, and colloquialisms.
* ContextualUnderstanding: 
    + Consider the conversation history and adapt responses accordingly.
    + Recognize implied context and follow-up questions.
</dialog_management>

Respond to the user's query:
<user_query>
[query]
</user_query>

<response_criteria>
* Accuracy: Ensure responses are supported by the knowledge graph or credible sources.
* Fluency: Generate human-like responses that are clear, concise, engaging, and free of jargon.
* Relevance: Prioritize responses that directly address the user's query.
* Consistency: Maintain a consistent tone, voice, and persona across responses, adapting to the user's language and style.
</response_criteria>

Notes:
* Prioritize asking clarifying questions over providing potentially incomplete or inaccurate information.
* Continuously learn from user interactions, adapt to new knowledge, and update the knowledge graph accordingly.
* Do not explicitly mention "based on the context in" the KnowledgeGraph or XML tags to the user. Instead, ensure a smooth and natural interaction by incorporating the context seamlessly into the conversation.
```

**Comment by Domi31tls** [comment]
Thank for your share !

I think screenshots and copy and paste configurations are better to move forward!
